### PR TITLE
Add ruff - Replace Black, isort and pyupgrade

### PR DESCRIPTION
**Pull request by mirpedrol**
Replace PR https://github.com/nf-core/tools/pull/2174
Close https://github.com/nf-core/tools/issues/2533

Implement [Ruff linter](https://github.com/astral-sh/ruff). In replacement of Black, isort and pyupgrade. It also includes Flake8.

### CODE REVIEW

File: .gitpod.yml [hunk 1]
```diff
@@ -21,3 +21,4 @@ vscode:
     # - nextflow.nextflow                    # Nextflow syntax highlighting
     - oderwat.indent-rainbow # Highlight indentation level
     - streetsidesoftware.code-spell-checker # Spelling checker for source code
+    - - charliermarsh.ruff # Code linter Ruff
```
Comment (mirpedrol) [inline .gitpod.yml:24]:
```suggestion
    - charliermarsh.ruff # Code linter Ruff
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,8 @@
 
 ### General
 
+- Add Ruff linter and formatter
```
Comment (mirpedrol) [inline CHANGELOG.md:17]:
```suggestion
- Add Ruff linter and formatter replacing Blac, isort and pyupgrade ([#2620](https://github.com/nf-core/tools/pull/2620))
```

File: nf_core/params_file.py [hunk 3]
```diff
@@ -197,7 +195,6 @@ def format_param(self, name, properties, required_properties=(), show_hidden=Fal
         self.schema_obj.get_schema_defaults()
         default = properties.get("default")
         typ = properties.get("type")
-        required = name in required_properties
```
Comment (mirpedrol) [inline nf_core/params_file.py:200]: should we use it?

File: nf_core/params_file.py [hunk 4]
```diff
@@ -195,6 +195,7 @@ def format_param(self, name, properties, required_properties=(), show_hidden=Fal
         self.schema_obj.get_schema_defaults()
         default = properties.get("default")
         typ = properties.get("type")
```
Comment (mashehu) [inline nf_core/params_file.py:197]: that looks like a typ_o 🙂

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -35,45 +35,28 @@ pip install -e .
...
-You can run Black on the command line (it's included in `requirements-dev.txt`) - eg. to run recursively on the whole repository:
+You can run Ruff on the command line (it's included in `requirements-dev.txt`) - eg. to run recursively on the whole repository:
 
 ```bash
-black .
+ruff format .
 ```
 
-Alternatively, Black has [integrations for most common editors](https://black.readthedocs.io/en/stable/editor_integration.html)
+Alternatively, Black has [integrations for most common editors](https://github.com/astral-sh/ruff-lsp) and VSCode(https://github.com/astral-sh/ruff-vscode)
```
Comment (mashehu) [inline .github/CONTRIBUTING.md:49]:
```suggestion
Alternatively, Ruff has [integrations for most common editors](https://github.com/astral-sh/ruff-lsp) and VSCode(https://github.com/astral-sh/ruff-vscode)
```

File: .github/workflows/create-lint-wf.yml [hunk 6]
```diff
@@ -76,6 +81,9 @@ jobs:
       - name: Run Prettier --check
         run: prettier --check create-lint-wf/nf-core-testpipeline
 
+      - name: Run Ruff check
+        run: ruff check create-lint-wf/nf-core-testpipeline
+
```
Comment (mashehu) [inline .github/workflows/create-lint-wf.yml:86]:
```suggestion
      - name: Run Ruff format
        run: ruff format create-lint-wf/nf-core-testpipeline
```

File: .github/workflows/lint-code.yml [hunk 7]
```diff
@@ -44,52 +44,45 @@ jobs:
...
+      - name: Set up Python 3.11
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.11
+      - name: Install Ruff
+        run: |
+          python -m pip install --upgrade pip
+          pip install ruff
+      - name: Run Ruff
+        run: ruff check .
```
Comment (mashehu) [inline .github/workflows/lint-code.yml:62]:
```suggestion
        run: ruff check .
        
     - name: Run Ruff check
        run: ruff check create-lint-wf/nf-core-testpipeline
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -14,6 +14,8 @@
 
 ### General
 
+- Add Ruff linter and formatter replacing Blac, isort and pyupgrade ([#2620](https://github.com/nf-core/tools/pull/2620))
```
Comment (mashehu) [inline CHANGELOG.md:17]:
```suggestion
- Add Ruff linter and formatter replacing Black, isort and pyupgrade ([#2620](https://github.com/nf-core/tools/pull/2620))
```

File: nf_core/pipeline-template/.github/workflows/linting.yml [hunk 9]
```diff
@@ -37,31 +37,40 @@ jobs:
...
+      - name: Set up Python 3.11
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.11
+      - name: Install Ruff
+        run: |
+          python -m pip install --upgrade pip
+          pip install ruff
+      - name: Run Ruff
+        run: ruff check .
```
Comment (mashehu) [inline nf_core/pipeline-template/.github/workflows/linting.yml:55]:
```suggestion
        run: ruff check .
     - name: Run Ruff check
        run: ruff check create-lint-wf/nf-core-testpipeline
```

File: nf_core/pipeline-template/pyproject.toml [hunk 10]
```diff
@@ -1,10 +1,14 @@
-# Config file for Python. Mostly used to configure linting of bin/check_samplesheet.py with Black.
+# Config file for Python. Mostly used to configure linting of bin/check_samplesheet.py with Ruff.
 # Should be kept the same as nf-core/tools to avoid fighting with template synchronisation.
-[tool.black]
+[tool.ruff]
 line-length = 120
```
Comment (mashehu) [inline nf_core/pipeline-template/pyproject.toml:4]: for this to have an effect, I think we need to select also rule `E5` https://docs.astral.sh/ruff/rules/line-too-long/

File: nf_core/pipeline-template/pyproject.toml [hunk 11]
```diff
@@ -1,10 +1,14 @@
...
-target_version = ["py37", "py38", "py39", "py310"]
+target-version = "py38"
+ignore = ["E501"]
+select = ["I", "E4", "E7", "E9", "F"]
```
Comment (mashehu) [inline nf_core/pipeline-template/pyproject.toml:7]: I think "E1" and "E2" would also make sense to have. Or would you like to do that in a separate PR, because it might include more changes (we can ignore that specific commit with ` --ignore-rev`)

Comment (mashehu) [reply nf_core/pipeline-template/pyproject.toml:7]: ["N"](https://docs.astral.sh/ruff/rules/#pep8-naming-n) for keeping variable and function names in line wuld be nice as well.

Comment (mashehu) [inline nf_core/pipeline-template/pyproject.toml:7]:
```suggestion
select = ["I", "E4", "E7", "E9", "F", "UP"]
```
To use pyupgrade

**Review by mashehu — APPROVED** [review]

File: nf_core/pipeline-template/pyproject.toml (hunk 11 above)
Comment (mashehu) [reply nf_core/pipeline-template/pyproject.toml:7]: see #2623

Comment (mirpedrol) [reply nf_core/pipeline-template/pyproject.toml]: I have added `E1`, `N` and `UP` to this PR, the changes were easy. `E2` is still unstable, and we see the message `warning: Selection E2 has no effect because the --preview flag was not included.`, so it will be better to add in once it is stable.

File: .github/workflows/lint-code.yml (hunk 7 above)
Comment (mirpedrol) [reply .github/workflows/lint-code.yml:62]: This folder doesn't exist here, did you mean to run `ruff format`?

Comment (mashehu) [reply .github/workflows/lint-code.yml:62]: sorry yes

Comment (mirpedrol) [reply .github/workflows/lint-code.yml:62]:
```suggestion
        run: ruff check .

     - name: Run Ruff format
        run: ruff format .
```

File: nf_core/pipeline-template/.github/workflows/linting.yml (hunk 9 above)
Comment (mirpedrol) [reply nf_core/pipeline-template/.github/workflows/linting.yml:55]:
```suggestion
        run: ruff check .

     - name: Run Ruff format
        run: ruff format .
```
